Introduce wl-summary-descending-order

shingo256 opened the pull request:
Hi Ito-san,

First of all, thank you for maintaining Wanderlust which is still awesome Emacs MUA as of today.  I would like to suggest to introduce wl-summary-descending-order to make Wanderlust a little bit align with modern mailers like Gmail, MS Outlook and so on.  This is a higher level concept than reverse sorting.  Meaning, reverse sorting when wl-summary-descending-order t, messages are listed in ascending order.  Hope this makes sense.

ikazuhiro commented:
Thank you for your pull request.

1. The patch for `wl-summary-sync-update` function seems to affect the behavior in the case that `wl-summary-descending-order` is `nil`.  Is it intended?  If so, why is it needed?

When `wl-summary-descending-order` is `t` and threading is on (`wl-summary-toggle-thread` on Summary buffer), 

2. `wl-summary-read` (`SPC` on Summary buffer) always moves cursor in reverse number order, i.e. thread order is ignored.  For example, when the first message (number 1) is the root of the thread and is displayed, you exit from Summary buffer by SPC and child messages are not displayed.

3. Thread's child messages are sorted in ascending order.  wl-summary-sort (`S` on Summary buffer) sorts also thread's child messages in specified order.  These behaviors are inconsistent.

shingo256 commented:
Hi Ito-san

Thank you for the advice.  Much appreciated.

> The patch for wl-summary-sync-update function seems to affect the behavior in the case that wl-summary-descending-order is nil. Is it intended? If so, why is it needed?

Sorry this was my lack of understanding.  Now I've fixed it.

> wl-summary-read (SPC on Summary buffer) always moves cursor in reverse number order, i.e. thread order is ignored. For example, when the first message (number 1) is the root of the thread and is displayed, you exit from Summary buffer by SPC and child messages are not displayed.

I've fixed it by reversing wl-thread-entity-list and wl-summary-buffer-number-list rather than reversing only displayed order, so that we can get consistent behavior between displayed order and wl-summary-read order.

> Thread's child messages are sorted in ascending order. wl-summary-sort (S on Summary buffer) sorts also thread's child messages in specified order. These behaviors are inconsistent.

The 1st sentence "Thread's child messages are sorted in ascending order" is the intended behavior as this is the modern mailers' behavior (Gmail, MS Outlook).  I've also added the explanation to wl-summary-descending-order doc for clarification.  So, when wl-summary-descending-order is t:

- wl-summary-sort sorts in order (i.e. "date") -> threads are descending, messages in each thread are ascending
- wl-summary-sort sorts in reverse order (i.e. "!date") -> threads are ascending, messages in each thread are ascending

This is the intended behavior.  Correct me if I am missing your point.

I also added info docs.  Please check.

ikazuhiro commented:
The logic seems ok.

But if it is intended that messages in thread are in ascending order, please change variable name `wl-summary-descending-order` to `wl-summary-order` with below definition for someone who may want to add new sort order in the future.
And, please replace conditional branch `(if wl-summary-descending-order A B)` with `(if (eq wl-summary-order 'descending) A B)`.

```elisp
(defcustom wl-summary-order 'asecnding
  "Specify order of messages in summary buffer.
Note that messages in a thread are always listed in ascending order
even if this value is descending."
  :type '(choice '(const :tag "Ascending order" ascending)
		 '(const :tag "Decending order" descending))
  :group 'wl-summary)
```

shingo256 commented:
Thank you for the advice.  Done.  Please check.

ikazuhiro commented:
Merged with slight change (NEWS entry for old version should not be modified, so reverted), thank you.
If you have the next time, please write commit log like ChangeLog entry.  See recent commit log of Wanderlust.
